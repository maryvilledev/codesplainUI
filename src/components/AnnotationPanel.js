import React, { PropTypes } from 'react';

import AnnotationDisplay from './AnnotationDisplay';
import AnnotationEditor from './AnnotationEditor';
import LineSnippet from './LineSnippet';

class AnnotationPanel extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isEditing: props.annotation.length === 0,
    };
    this.handleSaveAnnotation = this.handleSaveAnnotation.bind(this);
    this.toggleEditState = this.toggleEditState.bind(this);
  }

  toggleEditState() {
    this.setState({
      isEditing: !this.state.isEditing,
    });
  }

  handleSaveAnnotation(annotation) {
    this.props.saveAnnotation(annotation);
    this.toggleEditState();
  }

  render() {
    const {
      annotation,
      closeAnnotation,
      snippetInformation: {
        lineNumber,
        lineText,
      }
    } = this.props;
    const { isEditing } = this.state;
    const shouldRenderEditor = annotation.length === 0 || isEditing;
    return (
      <div>
        <LineSnippet
          lineNumber={lineNumber + 1}
          value={lineText}
        />
        { shouldRenderEditor ?
          <AnnotationEditor
            annotation={annotation}
            closeAnnotation={closeAnnotation}
            saveAnnotation={this.handleSaveAnnotation}
          /> :
          <AnnotationDisplay
            annotation={annotation}
            closeAnnotation={closeAnnotation}
            editAnnotation={this.toggleEditState}
          />
        }
      </div>
    );
  }

}

AnnotationPanel.propTypes = {
  annotation: PropTypes.string.isRequired,
  saveAnnotation: PropTypes.func.isRequired,
  closeAnnotation: PropTypes.func.isRequired,
  snippetInformation: PropTypes.shape({
    lineNumber: PropTypes.number.isRequired,
    lineText: PropTypes.string.isRequired,
  }).isRequired,
};

export default AnnotationPanel;
