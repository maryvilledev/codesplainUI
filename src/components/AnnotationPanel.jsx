import React, { PropTypes } from 'react';

import AnnotationDisplay from './AnnotationDisplay';
import AnnotationEditor from './AnnotationEditor';
import LineSnippet from './LineSnippet';

class AnnotationPanel extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isEditing: true,
    };
    this.handleSaveAnnotation = this.handleSaveAnnotation.bind(this);
    this.toggleEditState = this.toggleEditState.bind(this);
  }

  handleSaveAnnotation(annotation) {
    this.props.saveAnnotation(annotation);
    this.toggleEditState();
  }

  toggleEditState() {
    this.setState({
      isEditing: !this.state.isEditing,
    });
  }

  render() {
    const {
      annotation,
      closeAnnotation,
      snippetInformation: {
        lineNumber,
        lineText,
      },
    } = this.props;
    const { isEditing } = this.state;
    // AnnotationEditor should be rendered in two cases:
    // 1. When the annotation is empty
    // 2. When the state is explicitly set to edit mode
    const shouldRenderEditor = annotation.length === 0 || isEditing;
    return (
      <div>
        <LineSnippet
          lineNumber={lineNumber + 1}
          value={lineText}
        />
        { shouldRenderEditor ?
          <AnnotationEditor
            annotation={annotation}
            closeAnnotation={closeAnnotation}
            saveAnnotation={this.handleSaveAnnotation}
          /> :
          <AnnotationDisplay
            annotation={annotation}
            closeAnnotation={closeAnnotation}
            editAnnotation={this.toggleEditState}
          />
        }
      </div>
    );
  }

}

AnnotationPanel.propTypes = {
  annotation: PropTypes.string.isRequired,
  closeAnnotation: PropTypes.func.isRequired,
  saveAnnotation: PropTypes.func.isRequired,
  snippetInformation: PropTypes.shape({
    lineNumber: PropTypes.number.isRequired,
    lineText: PropTypes.string.isRequired,
  }).isRequired,
};

export default AnnotationPanel;
