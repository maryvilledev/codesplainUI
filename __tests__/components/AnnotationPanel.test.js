import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import getMuiTheme from 'material-ui/styles/getMuiTheme';

import AnnotationPanel from '../../src/components/AnnotationPanel';

const mockFunctionProps = {
  closeAnnotation: jest.fn(),
  saveAnnotation: jest.fn(),
};

describe('<AnnotationPanel />', () => {
  const muiTheme = getMuiTheme();
  const shallowWithContext = (node) => shallow(node, { context: { muiTheme } });

  describe('snapshot tests', () => {
    it('matches snapshot when no annotation is saved for a line', () => {
      const wrapper = shallowWithContext(
        <AnnotationPanel
          annotation=''
          snippetInformation={{
            lineNumber: 0,
            lineText: 'True',
          }}
          {...mockFunctionProps}
        />
      );
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });

    it('matches snapshot when an annotation is saved for a line', () => {
      const wrapper = shallowWithContext(
        <AnnotationPanel
          annotation='Wubba lubba dub dub!'
          snippetInformation={{
            lineNumber: 0,
            lineText: 'True',
          }}
          {...mockFunctionProps}
        />
      );
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
  });

  describe('prop: snippetInformation', () => {
    it('is forwarded to LineSnippet', () => {
      const lineInfo = {
        lineNumber: 0,
        lineText: 'You pass butter',
      };
      const wrapper = shallowWithContext(
        <AnnotationPanel
          annotation='some annotation'
          snippetInformation={lineInfo}
          {...mockFunctionProps}
        />
      );
      const lineSnippet = wrapper.find('LineSnippet');
      expect(lineSnippet.prop('lineNumber')).toEqual(Number(lineInfo['lineNumber']) + 1);
      expect(lineSnippet.prop('value')).toEqual(lineInfo['lineText']);
    });
  });
});
