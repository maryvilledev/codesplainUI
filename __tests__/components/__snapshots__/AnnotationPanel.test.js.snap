exports[`<AnnotationPanel /> snapshot tests matches snapshot when an annotation is saved for a line 1`] = `
<div>
  <LineSnippet
    lineNumber={1}
    value="True" />
  <AnnotationEditor
    annotation="Wubba lubba dub dub!"
    closeAnnotation={[Function]}
    saveAnnotation={[Function]} />
</div>
`;

exports[`<AnnotationPanel /> snapshot tests matches snapshot when no annotation is saved for a line 1`] = `
<div>
  <LineSnippet
    lineNumber={1}
    value="True" />
  <AnnotationEditor
    annotation=""
    closeAnnotation={[Function]}
    saveAnnotation={[Function]} />
</div>
`;
